warehouse/oidc/services.py

```diff
@@ -287,11 +310,30 @@ def find_publisher(
             publisher = find_publisher_by_issuer(
                 self.db, self.issuer_url, signed_claims, pending=pending
             )
+
+            jwt_token_identifier: str | None = signed_claims.get("jti")
+            # jti is in the __preverified_claims__ set, so if it was present,
+            # it was already checked
+            if pending is False and jwt_token_identifier:
+                if self.token_identifier_exists(jwt_token_identifier):
+                    self.metrics.increment(
+                        "warehouse.oidc.reused_token",
+                        tags=metrics_tags,
+                    )
+                    raise ReusedTokenError("JWT Token already used to mint a token.")
+
             publisher.verify_claims(signed_claims)
             self.metrics.increment(
                 "warehouse.oidc.find_publisher.ok",
                 tags=metrics_tags,
             )
+
+            if pending is False and jwt_token_identifier:
+                # Of note, exp is coming from a trusted source here,
+                # so we don't validate it
+                expiration = typing.cast(int, signed_claims.get("exp"))
+                self.store_jwt_identifier(jwt_token_identifier, expiration)
```

Comment:
Can we merge these two conditional blocks? I think the `store_jwt_identifier` could happen as a fall-through if `token_identifier_exists` returns `False`.

Reply:
I may have misunderstood your suggestion but the reasoning here was :

- Check if the identifier exists and bail out if it does
- Verify the claims (and return early if they are not verified)
- Finally, mark the JTI as already used

We only store JTI associated to valid claims to prevent flooding the database with signed but invalid claims.

Merging the two blocks would require to follow one of this two sequences : 

- Validating the uniqueness of the JTI, storing the JTI, validating the claims ( and maybe remove it if the claims are invalids)
- Validating the claims, validating the uniqueness of the JTI, and storing it.

If believe the current solution is more elegant, but introduces a slight duplication of conditions. However, I'll be happy to change it if that makes more sense.

Reply:
Oh, you're right -- I missed that `verify_claims` is called between the two blocks. This LGTM then 🙂